threadpoolctl.py

```diff
@@ -599,6 +211,520 @@ def get_original_num_threads(self):
+            libc = ctypes.CDLL(libc_name, mode=_RTLD_NOLOAD)
+            cls._system_libraries["libc"] = libc
+        return libc
+
+    @classmethod
+    def _get_windll(cls, dll_name):
+        """Load a windows DLL"""
+        dll = cls._system_libraries.get(dll_name)
+        if dll is None:
+            dll = ctypes.WinDLL("{}.dll".format(dll_name))
+            cls._system_libraries[dll_name] = dll
+        return dll
+
+    @classmethod
+    def _realpath(cls, filepath, cache_limit=10000):
+        """Small caching wrapper around os.path.realpath to limit system calls
+        """
+        rpath = cls._realpaths.get(filepath)
+        if rpath is None:
+            rpath = os.path.realpath(filepath)
+            if len(cls._realpaths) < cache_limit:
+                # If we drop support for Python 2.7, we could use
+                # functools.lru_cache with maxsize=10000 instead.
+                cls._realpaths[filepath] = rpath
+        return rpath
+
+
+class _Module(ABC):
+    """Abstract base class for the modules
+
+    A module is represented by the following information:
+      - "user_api" : user API. Possible values are {USER_APIS}.
+      - "internal_api" : internal API. Possible values are {INTERNAL_APIS}.
+      - "prefix" : prefix of the shared library's name.
+      - "filepath" : path to the loaded module.
+      - "version" : version of the library (if available).
+      - "num_threads" : the current thread limit.
+
+    In addition, each module may contain internal_api specific entries.
+    """
+    def __init__(self, filepath=None, prefix=None):
+        self.filepath = filepath
+        self.prefix = prefix
+        self.dynlib = ctypes.CDLL(filepath, mode=_RTLD_NOLOAD)
+        self.version = self.get_version()
+        self.num_threads = self.get_num_threads()
+        self._get_extra_info()
+
+    @property
+    @abstractmethod
+    def user_api(self):
+        pass  # pragma: no cover
+
+    @property
+    @abstractmethod
+    def internal_api(self):
+        pass  # pragma: no cover
+
+    def __eq__(self, other):
+        return self.todict() == other.todict()
+
+    def todict(self):
+        """Return relevant info wrapped in a dict"""
+        return {"user_api": self.user_api,
+                "internal_api": self.internal_api,
+                "prefix": self.prefix,
+                "filepath": self.filepath,
+                "version": self.version,
+                "num_threads": self.num_threads}
+
+    @abstractmethod
+    def get_version(self):
+        """Return the version of the shared library"""
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def get_num_threads(self):
+        """Return the maximum number of threads available to use"""
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def set_num_threads(self, num_threads):
+        """Set the maximum number of threads to use"""
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def _get_extra_info(self):
```

Comment:
Why is this abstract method private if all others are "public". The whole class is object oriented module API is private anyway so I think it would be fine to remove the leading `_`? WDYT?

Reply:
My policy for _method in a private class is a method intended to be used exclusively inside the class (a private method).